api/python/t4/snapshots.py

```diff
@@ -131,10 +139,345 @@ def download_file_from_snapshot(src, dst, snapshothash):
+        Gets objects from logical_key inside the package and saves them to path.
+
+        Args:
+            logical_key: logical key inside package to get
+            path: where to put the files
+
+        Returns:
+            None
+
+        Raises:
+            logical key not found
+            physical key failure
+            fail to create file
+            fail to finish write
+        """
+        raise NotImplementedError
+
+    def get_meta(self, logical_key):
+        """
+        Returns user metadata for specified logical key.
+        """
+        entry = self._data[logical_key]
+        return entry['user_meta']
+
+    def dump(self, writable_file):
+        """
+        Serializes this package to a writable file-like object.
+
+        Args:
+            writable_file: file-like object to write serialized package.
+
+        Returns:
+            None
+
+        Raises:
+            fail to create file
+            fail to finish write
+        """
+        writer = jsonlines.Writer(writable_file)
+        writer.write(self._meta)
+        for logical_key, entry in self._data.items():
+            writer.write({'logical_key': logical_key, **entry.as_dict()})
+
+    def update(self, logical_key, entry):
+        """
+        Returns a new package with the object at logical_key set to entry.
+
+        Args:
+            logical_key(string): logical key to update
+            entry(PackageEntry): new entry to place at logical_key in the package
+
+        Returns:
+            A new package
+        """
+        pkg = self._clone()
+        # TODO validate entry contents
+        pkg._data[logical_key] = entry
+        return pkg
+
+    def delete(self, logical_key):
+        """
+        Returns a new package with logical_key removed.
+
+        Returns:
+            A new package
+
+        Raises:
+            KeyError: when logical_key is not present to be deleted
+        """
+        pkg = self._clone()
+        pkg._data.pop(logical_key)
+        return pkg
+
+    def _top_hash(self):
+        """
+        Sets the top_hash in _meta
+
+        Returns:
+            None
+        """
+        raise NotImplementedError
+
+    def top_hash(self):
+        """
+        Returns the top hash of the package.
+
+        Note that physical keys are not hashed because the package has
+            the same semantics regardless of where the bytes come from.
+
+        Returns:
+            A string that represents the top hash of the package
+        """
+        if 'top_hash' not in self._meta:
+            self._top_hash()
+        return self._meta['top_hash']
+
+    def textual_hash(self):
+        """
+        Returns the textual hash of the package manifest serialized to disk
+
+        Returns:
+            SHA256 hash of serialized manifest
+        """
+        with tempfile.NamedTemporaryFile() as tmpfile:
+            tmp_name = tmpfile.name
+            self.dump(tmpfile)
+            self_hash = hash_file(tmp_name)
```

Comment:
Turns out, this will have the same problem as earlier: you're now opening the temp file for the second time.